executor/src/main/java/com/redhat/service/smartevents/executor/ExecutorImpl.java

```diff
@@ -0,0 +1,121 @@
+import com.redhat.service.smartevents.infra.models.dto.ProcessorDTO;
+import com.redhat.service.smartevents.infra.models.gateways.Action;
+import com.redhat.service.smartevents.infra.transformations.TransformationEvaluator;
+import com.redhat.service.smartevents.infra.transformations.TransformationEvaluatorFactory;
+import com.redhat.service.smartevents.infra.utils.CloudEventUtils;
+import com.redhat.service.smartevents.processor.actions.ActionInvoker;
+import com.redhat.service.smartevents.processor.actions.ActionRuntime;
+
+import io.cloudevents.CloudEvent;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+
+public class ExecutorImpl implements Executor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ExecutorImpl.class);
+
+    private final ProcessorDTO processor;
+    private final FilterEvaluator filterEvaluator;
+    private final TransformationEvaluator transformationEvaluator;
+    private final ActionInvoker actionInvoker;
+
+    private Timer processorProcessingTime;
+    private Timer filterTimer;
+    private Timer actionTimer;
+    private Timer transformationTimer;
+
+    public ExecutorImpl(
+            ProcessorDTO processor,
+            FilterEvaluatorFactory filterEvaluatorFactory,
+            TransformationEvaluatorFactory transformationFactory,
+            ActionRuntime actionRuntime,
+            MeterRegistry registry) {
+        this.processor = processor;
+        this.filterEvaluator = filterEvaluatorFactory.build(processor.getDefinition().getFilters());
+        this.transformationEvaluator = transformationFactory.build(processor.getDefinition().getTransformationTemplate());
+
+        Action action = processor.getDefinition().getResolvedAction();
+        this.actionInvoker = actionRuntime.getInvokerBuilder(action.getType()).build(processor, action);
+
+        initMetricFields(processor, registry);
+    }
+
+    @Override
+    public ProcessorDTO getProcessor() {
+        return processor;
+    }
+
+    @Override
+    public void onEvent(CloudEvent event) {
+        processorProcessingTime.record(() -> process(event));
+    }
+
+    private void process(CloudEvent event) {
+        Map<String, Object> eventMap = toEventMap(event);
+
+        LOG.debug("Received event with id '{}' and type '{}' in processor with name '{}' of bridge '{}", event.getId(), event.getType(), processor.getName(), processor.getBridgeId());
+
+        // Filter evaluation
+        if (!matchesFilters(eventMap)) {
+            LOG.debug("Filters of processor '{}' did not match for event with id '{}' and type '{}'", processor.getId(), event.getId(), event.getType());
+            return;
+        }
+        LOG.info("Filters of processor '{}' matched for event with id '{}' and type '{}'", processor.getId(), event.getId(), event.getType());
+        // Transformation
+        String eventToSend = applyTransformations(eventMap);
```

Comment:
@kostola For a `Source` processor, how do we guarantee that this transforms into a `CloudEvent`? From what I can see we simply send whatever the customer transforms into a JSON payload and send it to the Ingress via Webhook Action. Am I correct?

I think we have two options here:

- If the Processor is of type `Source`, then always marshall the transformed event back into a CloudEvent format
- Prevent the user from applying a transformation to a `Source` processor. As discussed before, that is out-of-scope for this work.

cheers,

Rob

Reply:
There's no guarantee that the transformed event will always be a `CloudEvent` with the current implementation.

I refactored the code to include your second suggestion, which is also what I believe is the best solution at the moment:

> Prevent the user from applying a transformation to a Source processor

This allows us to finalize the first version of Event Sources quickly and think more in depth about transformations later. I honestly think source processors feature can initially be merged like this without compromising its usage.

Reply:
@kostola I agree with you. I've created https://issues.redhat.com/browse/MGDOBR-682 to capture the need for `Transformation` on an Event Source.

cheers,
Rob